modules/sor/lib/pathGraph/pathGraph.ts

```diff
@@ -86,25 +86,47 @@ export class PathGraph {
         const paths: PathGraphEdgeData[][] = [];
         const selectedPathIds: string[] = [];
 
-        // the tokenPairIndex refers to the nth most liquid path for a token
-        // pair x -> y. maxPathsPerTokenPair is provided as a config on graph init
-        for (let idx = 0; idx < this.maxPathsPerTokenPair; idx++) {
-            for (let i = 0; i < tokenPaths.length; i++) {
-                const path = this.expandTokenPath({
-                    tokenPath: tokenPaths[i],
-                    tokenPairIndex: idx,
-                });
+        // For each token path, explore all possible combinations of liquidity ranks
+        for (let i = 0; i < tokenPaths.length; i++) {
+            const tokenPath = tokenPaths[i];
+            const pathLength = tokenPath.length - 1; // number of segments in the path
+
+            // Generate all possible combinations of liquidity ranks for this path
+            // Each element is an array of ranks, one for each segment in the path
+            // For example: [[0,0], [0,1], [1,0], [1,1]] for a 2-segment path with maxPathsPerTokenPair=2
+            let rankCombinations: number[][] = [[]];
+            for (let segment = 0; segment < pathLength; segment++) {
+                const newCombinations: number[][] = [];
+                for (const combo of rankCombinations) {
+                    for (let rank = 0; rank < this.maxPathsPerTokenPair; rank++) {
+                        newCombinations.push([...combo, rank]);
+                    }
+                }
+                rankCombinations = newCombinations;
+            }
+
+            // Now iterate through all combinations and expand paths
+            for (const ranks of rankCombinations) {
+                try {
+                    const path = this.expandTokenPathWithRanks({
+                        tokenPath: tokenPath,
+                        ranks: ranks,
```

Comment:
This is the main change.
Instead of passing a single `tokenPairIndex` to the expandTokenPath logic, we're now passing an array of `ranks`.
This is done for all possible combinations within the `maxPathsPerTokenPair` limit.